lib/internal/streams/web.js

```diff
@@ -0,0 +1,4632 @@
+    const preventClose = options?.preventClose;
+    const signal = options?.signal;
+
+    if (signal !== undefined && signal?.[kAborted] === undefined)
+      throw new ERR_INVALID_ARG_TYPE('options.signal', 'AbortSignal', signal);
+
+    if (isReadableStreamLocked(this))
+      throw new ERR_INVALID_STATE.TypeError('The ReadableStream is locked');
+    if (isWritableStreamLocked(writable))
+      throw new ERR_INVALID_STATE.TypeError('The WritableStream is locked');
+
+    const promise = readableStreamPipeTo(
+      this,
+      writable,
+      !!(preventClose),
+      !!(preventAbort),
+      !!(preventCancel),
+      signal);
+    setPromiseHandled(promise);
+
+    return readable;
+  }
+
+  /**
+   * @param {WritableStream} destination
+   * @param {StreamPipeOptions} [options]
+   * @returns {Promise<void>}
+   */
+  pipeTo(destination, options = {}) {
+    try {
+      if (!isReadableStream(this))
+        throw new ERR_INVALID_THIS('ReadableStream');
+      if (!isWritableStream(destination)) {
+        throw new ERR_INVALID_ARG_TYPE(
+          'transform.writable',
+          'WritableStream',
+          destination);
+      }
+
+      const preventAbort = options?.preventAbort;
+      const preventCancel = options?.preventCancel;
+      const preventClose = options?.preventClose;
+      const signal = options?.signal;
+
+      if (signal !== undefined && signal?.[kAborted] === undefined)
+        throw new ERR_INVALID_ARG_TYPE('options.signal', 'AbortSignal', signal);
+
+      if (isReadableStreamLocked(this))
+        throw new ERR_INVALID_STATE.TypeError('The ReadableStream is locked');
+      if (isWritableStreamLocked(destination))
+        throw new ERR_INVALID_STATE.TypeError('The WritableStream is locked');
+
+      return readableStreamPipeTo(
+        this,
+        destination,
+        !!(preventClose),
+        !!(preventAbort),
+        !!(preventCancel),
+        signal);
+    } catch (error) {
+      return PromiseReject(error);
+    }
+  }
+
+  /**
+   * @returns {ReadableStream[]}
+   */
+  tee() {
+    if (!isReadableStream(this))
+      throw new ERR_INVALID_THIS('ReadableStream');
+    return readableStreamTee(this, false);
+  }
+
+  /**
+   * @param {{
+   *   preventCancel? : boolean,
+   * }} [options]
+   * @returns {AsyncIterable}
+   */
+  values(options = {}) {
+    if (!isReadableStream(this))
+      throw new ERR_INVALID_THIS('ReadableStream');
+    validateObject(options, 'options');
+    const {
+      preventCancel = false,
+    } = options;
+
+    const reader = new ReadableStreamDefaultReader(this);
+    let done = false;
+    let started = false;
+    let current;
+
+    function nextSteps() {
+      if (done)
+        return PromiseResolve({ done: true, value: undefined });
+
+      if (reader[kState].stream === undefined) {
+        return PromiseReject(
+          new ERR_INVALID_STATE.TypeError(
+            'The reader is not bound to a ReadableStream'));
+      }
```

Comment:
```suggestion
    async function nextSteps() {
      if (done)
        return { done: true, value: undefined };

      if (reader[kState].stream === undefined) {
        throw new ERR_INVALID_STATE.TypeError(
            'The reader is not bound to a ReadableStream');
      }
```

Reply:
I'd rather not switch this to an async function. Given that it creates a Promise explicitly that is driven by the read request object and returns that, wrapping this in an async function ends up allocating additional unnecessary Promises that just add cost.

Reply:
v8 has lots of optimizations that prevent unnecessary promises from being created when using async/await.

Reply:
Take this script:

```js
const { createHook } = require('async_hooks');
let count = 0;
const hook = createHook({
  init(id, type) {
    if (type === 'PROMISE')
      count++;
  }
});

hook.enable();

process.on('exit', () => console.log(count));

function foo() {
  const p = new Promise(() => {});
  return p;
}

async function foo2() {
  const p = new Promise(() => {});
  return p;
}

eval(`${process.argv[2]}()`)
```

Running foo() creates 1 Promise.
Running foo2() creates 3.

```
root@DESKTOP-5KK9VIR:~/node/node# ./node ../tmp/a foo
1
root@DESKTOP-5KK9VIR:~/node/node# ./node ../tmp/a foo2
3
root@DESKTOP-5KK9VIR:~/node/node#
```

Reply:
The additional promises created here also change the timing of the microtask queue completions which throw off several of the web platform tests as currently written.

Reply:
For now, I'm going to leave this as is, without converting it to an async function.